### PR TITLE
Add LlamaForSequenceClassification

### PR DESCRIPTION
**Pull request by lewtun**
# What does this PR do?

This PR adds the `LlamaForSequenceClassification` class, which among standard applications can be used for reward modelling in the RLHF pipeline :)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc @ArthurZucker and @younesbelkada

### CODE REVIEW

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -831,3 +824,105 @@ def _reorder_cache(past_key_values, beam_idx):
         for layer_past in past_key_values:
             reordered_past += (tuple(past_state.index_select(0, beam_idx) for past_state in layer_past),)
         return reordered_past
+
+
+class LlamaForSequenceClassification(LlamaPreTrainedModel):
+    _keys_to_ignore_on_load_missing = [r"lm_head.weight"]
```
Comment (lewtun) [inline src/transformers/models/llama/modeling_llama.py:830]: I'm not entirely sure which other keys should be ignored for the LLaMa implementation - some guidance here is appreciated!

Comment (younesbelkada) [reply src/transformers/models/llama/modeling_llama.py:830]: I think that it should be similar to the ones in `LlamaForCausalLM`

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for this addition! I just left some small nits 🚀

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -831,3 +824,105 @@ def _reorder_cache(past_key_values, beam_idx):
...
+
+        # Model parallel
+        self.model_parallel = False
+        self.device_map = None
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    @add_start_docstrings_to_model_forward(LLAMA_INPUTS_DOCSTRING)
+    def forward(
```
[the remark this answers is not in the document]
Comment (lewtun) [reply src/transformers/models/llama/modeling_llama.py:844]: Hmm, it's not actually copied from there because LLaMa doesn't accept `token_type_ids` among other args. Happy to be corrected though if I'm wrong (I don't seem many other `# Copied from ...` statements in the llama modelling code)

File: src/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -831,3 +824,103 @@ def _reorder_cache(past_key_values, beam_idx):
...
+class LlamaForSequenceClassification(LlamaPreTrainedModel):
+    _keys_to_ignore_on_load_missing = [r"lm_head.weight"]
+
+    def __init__(self, config):
+        super().__init__(config)
+        self.num_labels = config.num_labels
+        self.transformer = LlamaModel(config)
+        self.score = nn.Linear(config.hidden_size, self.num_labels, bias=False)
+
+        # Model parallel
```
Comment (younesbelkada) [inline src/transformers/models/llama/modeling_llama.py:838]:
```suggestion
```

**Review by younesbelkada — APPROVED** [review]
LGTM with only one nit! 
I would also add potentially a test with `single_label_classification` to make sure everything works!
Let's wait for @ArthurZucker and/or @amyeroberts before merging!
Again thanks a lot for this

**Review by amyeroberts — APPROVED** [review]
Very nice 🤩 Thanks for adding!

File: src/transformers/models/llama/modeling_llama.py [hunk 4]
```diff
@@ -831,3 +824,103 @@ def _reorder_cache(past_key_values, beam_idx):
         for layer_past in past_key_values:
             reordered_past += (tuple(past_state.index_select(0, beam_idx) for past_state in layer_past),)
         return reordered_past
+
+
+class LlamaForSequenceClassification(LlamaPreTrainedModel):
```
[the remark this answers is not in the document]
Comment (lewtun) [reply src/transformers/models/llama/modeling_llama.py:844]: Done in https://github.com/huggingface/transformers/pull/22209/commits/c214270dba7edb01e5ded977fb83a1832bf0e67c :)

**Comment by lewtun** [comment]
> I would also add potentially a test with single_label_classification to make sure everything works!

Done in https://github.com/huggingface/transformers/pull/22209/commits/6737e380fc6a4cb73150da4fa821dd463f9a7204 :)

**Comment by younesbelkada** [comment]
Awesome thanks a lot @lewtun ! 🚀

File: src/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -831,3 +824,122 @@ def _reorder_cache(past_key_values, beam_idx):
...
+    def __init__(self, config):
+        super().__init__(config)
+        self.num_labels = config.num_labels
+        self.model = LlamaModel(config)
+        self.score = nn.Linear(config.hidden_size, self.num_labels, bias=False)
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_input_embeddings(self):
```
Comment (lewtun) [inline src/transformers/models/llama/modeling_llama.py:856]: FYI @younesbelkada @amyeroberts these getter/setters were missing in my original implementation, but now all tests pass locally for me and this should be good to go IMO :)

**Review by amyeroberts — APPROVED** [review]
❤️  - thanks for adding!
